### PR TITLE
fix: erase symbol from old symbol table when importing from a new module

### LINKED ISSUE #4165: bug: remove symbol from old parent symbol table when imported from a new module
### Minimum Reproducible Example

```fortran 
MODULE one
   IMPLICIT NONE
   INTEGER :: nx = 1
END MODULE one

MODULE two
   USE one, ONLY: nx
   IMPLICIT NONE
END MODULE two

PROGRAM test
   USE two
   USE one, ONLY: nx

   PRINT *, nx
END PROGRAM test
```
### Output
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
semantic error: Variable already defined
  --> ./examples/example.f90:13:4
   |
13 |    USE one, ONLY: nx
   |    ^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

This bug was caught when trying to compile the file `snap_main.f90` of the SNAP package. In there, `utils_module` defined inside file `utils.f90` imports the symbol `iproc` from `plib_module`. Inside the file `snap_main.f90`, `utils_module` is imported completely. After a few lines, some selected symbols are imported from `plib_module` which includes `iproc`. This leads to the above error. Another instance of this error occurs in the same file when importing `swp_typ` from `control_module` fails because `utils_module` already imports it.

GFortran does not throw an error here. While it is not very clear from the output of `-fdump-tree-all` how it handles this case, but I am guessing it might be removing the old symbol from the parent symbol table and re-declaring it in the new one.

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #4165 

```fortran
MODULE one
   IMPLICIT NONE
   INTEGER :: nx = 1
END MODULE one

MODULE two
   USE one, ONLY: nx
   IMPLICIT NONE
END MODULE two

PROGRAM test
   USE two
   USE one, ONLY: nx

   PRINT *, nx
END PROGRAM test
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
1
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
> This Error might be helpful in some cases. Also, we might sometimes overwrite the symbols.

Could you please give an example for this? I think you are talking about the case where we have a symbol of the same name declared in the program? As we import all symbols at the top, the imported symbols will be the first to enter the symbol table. So if we try to declare a symbol of the same name, the `declaration_util` will catch this error.

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -270,6 +270,8 @@ RUN(NAME include_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c fortran
 RUN(NAME include_02 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c fortran)
 RUN(NAME include_03 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c INCLUDE_PATH include_03 fortran)
 
+RUN(NAME use_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm c fortran)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:273]: The C test does not work for this.
```suggestion
RUN(NAME use_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc wasm fortran)
```

**Comment by certik** [comment]
Yes, we need to be very careful to ensure we are not removing valid error messages.

In Fortran, the `use` is always at the top, before all declarations. As long as the above symbol removal is always from the current scope (never from a parent scope), then I *think* the old symbol couldn't have been used.

The old symbol might be the same as the new symbol, but it might also be different. In both cases we need to shadow it.

However, we want to give a warning that the symbol is shadowed. I believe GFortran also gives a warning, at least in some cases (not sure only if the symbols are different or in all cases).

Let's do the following: give a warning in all the cases above when the old symbol is shadowed. That way if in the future we discover some issue with this, the warning will lead us to the right place in the code to fix.

File: integration_tests/CMakeLists.txt (hunk 1 above)
Comment (kmr-srbh) [reply integration_tests/CMakeLists.txt]: Yes, it was an issue with the C back-end. I have a PR open for this at #4199.

**Comment by kmr-srbh** [comment]
> As long as the above symbol removal is always from the current scope (never from a parent scope), then I think the old symbol couldn't have been used.

Yes, we remove the symbol from the current scope only.

**Comment by certik** [comment]
Go ahead and change the error to a warning, then we can merge this.

**Comment by kmr-srbh** [comment]
We now raise a warning.

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
warning: Symbol 'nx' from module 'one' shadows 'nx' in the current scope
  --> ./examples/example.f90:13:4
   |
13 |    USE one, ONLY: nx
   |    ^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
1
```

**Review by certik — APPROVED** [review]
Excellent, I think this is good now. The warning will show us if there are issues with this going forward.

**Comment by certik** [comment]
@kmr-srbh you have to remove the C test, since it's failing. I removed it above, but you added it back in. 

Let's remove it, get the CI to pass, merge this, and then merge the other PR. Then we can come back to the C test if we want to fix it.

**Comment by kmr-srbh** [comment]
> @kmr-srbh you have to remove the C test, since it's failing. I removed it above, but you added it back in.
> 
> Let's remove it, get the CI to pass, merge this, and then merge the other PR. Then we can come back to the C test if we want to fix it.

Oh, I am sorry. I thought you meant adding it back in https://github.com/lfortran/lfortran/pull/4199#issuecomment-2156714572.

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
Thanks! Now you can rebase your other PR(s).

**Pull request merged**